### DISCUSSION TITLE
RFC: add bambu lab printer support

### DISCUSSION THREAD

**Discussion by TimTheBig** [GladiusSlicer/GladiusSlicer#14] [Ideas]
I would like to add support for Bambu lab printers like my x1c.
Bambu lab printers implement most of [Marlin's commands](https://forum.bambulab.com/t/bambu-lab-x1-specific-g-code/666).

[Bambu lab specific g-code](https://forum.bambulab.com/t/bambu-lab-x1-specific-g-code/666/18)

**Comment by mrhatman** [comment]
The x1c appears to have multiple material support which this slicer does not yet. 

If you only need single material support then

The custom starting and ending instructions can be set via the starting_instructions and ending_instructions is the settings file.
` "starting_instructions":"M107;\nG90 ; use ........` https://github.com/GladiusSlicer/GladiusSlicer/blob/main/settings/PrusaMk3.json 

There appears to be some printer specific commands in the example like flush_length_1 and new_filament_e_feedrate. Uploading a full gcode that is know good would help identify what is unique about the printer and what is handled by the slicer. We support a few replacements that can be seen  here: https://github.com/GladiusSlicer/GladiusSlicer/blob/main/src/converter.rs#L296

**Comment by TimTheBig** [comment]
### Gcode Snippets From Orca Slicer
[Start gcode](https://pastebin.com/vUgQWDiX)
[End gcode](https://pastebin.com/3T12nTkT)
[Layer change gcode](https://pastebin.com/XqE1HsNG)

### A print with most of the body trimmed
[Full gcode](https://pastebin.com/7ezRCXD4)

**Reply by mrhatman** [reply]
Looking through some of these and others implementions, this should be fairly doable. 

Most slicers have a much better macro/constants system than we have so far.

I can mock something up later that should be able to do the job.

I need to think through how macros should work as the Slic3r implementation is a little clunky

**Comment by TimTheBig** [comment]
I'm adding bed_exclude_areas, were should I add the checks?

**Reply by mrhatman** [reply]
It will be needed to be added into the settings as an optional list of polygons, the check should be Part of check_model_ bounds in bound_checking.rs and return a new error variant if it fails

**Reply by TimTheBig** [reply]
I have implemented it [in this branch](https://github.com/TimTheBig/GladiusSlicer/tree/Bambu-Lab), how ever I still have some issues, with the start g-code being on multiple layers when it should be on one, so I will open a draft pull request once my previous patch is merged.
